File: android/app/src/main/java/com/grokremote/app/data/GrokRemoteClient.kt
package com.grokremote.app.data

import android.util.Log
import com.grokremote.app.model.ConnectionStatus
import com.grokremote.app.model.ServerProfile
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

class GrokRemoteClient {
	private companion object {
		const val LOG_TAG = "GrokRemoteClient"
	}
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    private val http = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .pingInterval(12, TimeUnit.SECONDS)
        .build()

    private val _connection = MutableStateFlow(ConnectionStatus.DISCONNECTED)
    val connection = _connection.asStateFlow()
    private val notificationQueue = Channel<JsonObject>(Channel.UNLIMITED)
    val notifications = notificationQueue.receiveAsFlow()

    private var profile: ServerProfile? = null
    private var webSocket: WebSocket? = null
    private var openSignal: CompletableDeferred<Unit>? = null
    private val ids = AtomicLong(0)
    private val pending = ConcurrentHashMap<Long, CompletableDeferred<JsonElement>>()
    private var closedByUser = false

    suspend fun connect(profile: ServerProfile): JsonObject {
        disconnect(notify = false)
        this.profile = profile
        closedByUser = false
        _connection.value = ConnectionStatus.CONNECTING

        val base = profile.baseUrl.toHttpUrl()
        val httpSocketUrl = base.newBuilder()
            .encodedPath("/ws")
            .query(null)
            .addQueryParameter("key", profile.key)
            .build()
            .toString()
        val wsUrl = if (httpSocketUrl.startsWith("https://")) {
            "wss://" + httpSocketUrl.removePrefix("https://")
        } else {
            "ws://" + httpSocketUrl.removePrefix("http://")
        }
        val request = Request.Builder()
            .url(wsUrl)
            .header("X-Grok-Remote-Key", profile.key)
            .build()
        val signal = CompletableDeferred<Unit>()
        openSignal = signal
        webSocket = http.newWebSocket(request, listener)
        withTimeout(20_000) { signal.await() }

        val result = rpc(
            method = "initialize",
            params = buildJsonObject {
                put("protocolVersion", 1)
                put("clientInfo", buildJsonObject {
                    put("name", "grok-remote-app-android")
                    put("version", "0.1.0")
                })
                put("clientCapabilities", buildJsonObject {
                    put("fs", buildJsonObject {
                        put("readTextFile", true)
                        put("writeTextFile", true)
                    })
                    put("terminal", true)
                })
            },
            timeoutMs = 20_000,
        )
        _connection.value = ConnectionStatus.CONNECTED
        return result as? JsonObject ?: JsonObject(emptyMap())
    }

    fun disconnect(notify: Boolean = true) {
        closedByUser = true
        webSocket?.close(1000, "client disconnect")
        webSocket = null
        openSignal?.cancel()
        openSignal = null
        failPending(IOException("connection closed"))
        if (notify) _connection.value = ConnectionStatus.DISCONNECTED
    }

    suspend fun rpc(method: String, params: JsonObject, timeoutMs: Long = 120_000): JsonElement {
        val socket = webSocket ?: error("连接已断开")
        val id = ids.incrementAndGet()
        val deferred = CompletableDeferred<JsonElement>()
        pending[id] = deferred
        val payload = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id)
            put("method", method)
            put("params", params)
        }
        if (!socket.send(payload.toString())) {
            pending.remove(id)
            error("WebSocket 发送失败")
        }
        return try {
            withTimeout(timeoutMs) { deferred.await() }
        } finally {
            pending.remove(id)
        }
    }

    fun notify(method: String, params: JsonObject) {
        val payload = buildJsonObject {
            put("jsonrpc", "2.0")
            put("method", method)
            put("params", params)
        }
        webSocket?.send(payload.toString())
    }

    fun reply(id: Long, result: JsonObject) {
        val payload = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id)
            put("result", result)
        }
        webSocket?.send(payload.toString())
    }

    suspend fun rest(
        path: String,
        method: String = "GET",
        query: Map<String, String> = emptyMap(),
        body: JsonObject? = null,
    ): JsonObject = withContext(Dispatchers.IO) {
        val current = profile ?: error("连接已断开")
        val builder = current.baseUrl.toHttpUrl().newBuilder().encodedPath(path).query(null)
        query.forEach { (key, value) -> builder.addQueryParameter(key, value) }
        val mediaType = "application/json; charset=utf-8".toMediaType()
        val requestBody = if (method == "GET" || method == "HEAD") null
        else (body ?: JsonObject(emptyMap())).toString().toRequestBody(mediaType)
        val request = Request.Builder()
            .url(builder.build())
            .header("X-Grok-Remote-Key", current.key)
            .header("Accept", "application/json")
            .method(method, requestBody)
            .build()
        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            val parsed = runCatching { json.parseToJsonElement(text).jsonObject }.getOrElse { JsonObject(emptyMap()) }
            if (!response.isSuccessful) {
                val message = (parsed["error"] as? JsonPrimitive)?.contentOrNull
                    ?: (parsed["message"] as? JsonPrimitive)?.contentOrNull
                    ?: "HTTP ${response.code}"
                error(message)
            }
            parsed
        }
    }

    fun close() {
        disconnect()
        notificationQueue.close()
        scope.cancel()
        http.dispatcher.executorService.shutdown()
    }

    private val listener = object : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
			Log.i(LOG_TAG, "WebSocket connected (${response.code})")
            openSignal?.complete(Unit)
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            val objectValue = runCatching { json.parseToJsonElement(text) as? JsonObject }.getOrNull() ?: return
            val id = (objectValue["id"] as? JsonPrimitive)?.longOrNull
            if (id != null && objectValue["method"] == null) {
                val deferred = pending.remove(id) ?: return
                val errorObject = objectValue["error"] as? JsonObject
                if (errorObject != null) {
                    val message = (errorObject["message"] as? JsonPrimitive)?.contentOrNull ?: "RPC error"
                    deferred.completeExceptionally(IOException(message))
                } else {
                    deferred.complete(objectValue["result"] ?: JsonNull)
                }
                return
            }
            notificationQueue.trySend(objectValue)
        }

        override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
            webSocket.close(code, reason)
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
			Log.w(LOG_TAG, "WebSocket closed: code=$code reason=$reason")
            this@GrokRemoteClient.webSocket = null
            failPending(IOException(reason.ifEmpty { "connection closed" }))
            if (!closedByUser) _connection.value = ConnectionStatus.RECONNECTING
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
			Log.e(LOG_TAG, "WebSocket failed: HTTP ${response?.code ?: "n/a"}", t)
            this@GrokRemoteClient.webSocket = null
            openSignal?.completeExceptionally(t)
            failPending(t)
            if (!closedByUser) _connection.value = ConnectionStatus.RECONNECTING
        }
    }

    private fun failPending(error: Throwable) {
        pending.values.forEach { it.completeExceptionally(error) }
        pending.clear()
    }
}

File: ios/GrokRemote/Core/GrokRemoteClient.swift
import Foundation

@MainActor
final class GrokRemoteClient {
    var onNotification: (([String: Any]) -> Void)?
    var onDisconnect: ((Error?) -> Void)?

    private var profile: ServerProfile?
    private var socket: URLSessionWebSocketTask?
    private var receiveTask: Task<Void, Never>?
    private var pending: [Int: CheckedContinuation<Any, Error>] = [:]
    private var nextID = 0

    var isConnected: Bool { socket?.state == .running }

    func connect(profile: ServerProfile) async throws -> [String: Any] {
        disconnect(notify: false)
        self.profile = profile

        var components = URLComponents(url: profile.baseURL, resolvingAgainstBaseURL: false)
        components?.scheme = profile.baseURL.scheme == "https" ? "wss" : "ws"
        components?.path = "/ws"
        components?.queryItems = [URLQueryItem(name: "key", value: profile.key)]
        guard let url = components?.url else { throw ClientError.invalidAddress }

        var request = URLRequest(url: url)
        request.setValue(profile.key, forHTTPHeaderField: "X-Grok-Remote-Key")
        let task = URLSession.shared.webSocketTask(with: request)
        socket = task
        task.resume()
        receiveTask = Task { [weak self] in await self?.receiveLoop() }

        let raw = try await rpc("initialize", params: [
            "protocolVersion": 1,
            "clientInfo": ["name": "grok-remote-app-ios", "version": "0.1.0"],
            "clientCapabilities": [:]
        ], timeout: 20)
        return raw as? [String: Any] ?? [:]
    }

    func disconnect(notify: Bool = true) {
        receiveTask?.cancel()
        receiveTask = nil
        socket?.cancel(with: .goingAway, reason: nil)
        socket = nil
        let continuations = Array(pending.values)
        pending.removeAll()
        continuations.forEach { $0.resume(throwing: ClientError.disconnected) }
        if notify { onDisconnect?(nil) }
    }

    func rpc(_ method: String, params: [String: Any], timeout: TimeInterval = 120) async throws -> Any {
        guard let socket, socket.state == .running else { throw ClientError.disconnected }
        nextID += 1
        let id = nextID
        let payload: [String: Any] = ["jsonrpc": "2.0", "id": id, "method": method, "params": params]

        return try await withCheckedThrowingContinuation { continuation in
            pending[id] = continuation
            Task { [weak self] in
                do {
                    try await self?.send(payload)
                    try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                    guard let self, let continuation = self.pending.removeValue(forKey: id) else { return }
                    continuation.resume(throwing: ClientError.server("请求超时：\(method)"))
                } catch {
                    guard let self, let continuation = self.pending.removeValue(forKey: id) else { return }
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    func reply(id: Int, result: [String: Any]) async throws {
        try await send(["jsonrpc": "2.0", "id": id, "result": result])
    }

    func notify(_ method: String, params: [String: Any]) async throws {
        try await send(["jsonrpc": "2.0", "method": method, "params": params])
    }

    func rest(path: String, method: String = "GET", query: [URLQueryItem] = [], body: [String: Any]? = nil) async throws -> [String: Any] {
        guard let profile else { throw ClientError.disconnected }
        guard var components = URLComponents(url: profile.baseURL, resolvingAgainstBaseURL: false) else {
            throw ClientError.invalidAddress
        }
        components.path = path
        components.queryItems = query.isEmpty ? nil : query
        guard let url = components.url else { throw ClientError.invalidAddress }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue(profile.key, forHTTPHeaderField: "X-Grok-Remote-Key")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw ClientError.malformedResponse }
        let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        guard (200..<300).contains(http.statusCode) else {
            throw ClientError.server(object.string("error", "message") ?? "HTTP \(http.statusCode)")
        }
        return object
    }

    private func send(_ payload: [String: Any]) async throws {
        guard let socket else { throw ClientError.disconnected }
        let data = try JSONSerialization.data(withJSONObject: payload)
        guard let text = String(data: data, encoding: .utf8) else { throw ClientError.malformedResponse }
        try await socket.send(.string(text))
    }

    private func receiveLoop() async {
        do {
            while !Task.isCancelled, let socket {
                let message = try await socket.receive()
                let data: Data
                switch message {
                case .string(let text): data = Data(text.utf8)
                case .data(let binary): data = binary
                @unknown default: continue
                }
                guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }
                handle(object)
            }
        } catch {
            guard !Task.isCancelled else { return }
            socket = nil
            let continuations = Array(pending.values)
            pending.removeAll()
            continuations.forEach { $0.resume(throwing: error) }
            onDisconnect?(error)
        }
    }

    private func handle(_ object: [String: Any]) {
        if let id = (object["id"] as? NSNumber)?.intValue,
           object["method"] == nil,
           let continuation = pending.removeValue(forKey: id) {
            if let error = object["error"] as? [String: Any] {
                continuation.resume(throwing: ClientError.server(error.string("message") ?? "RPC error"))
            } else {
                continuation.resume(returning: object["result"] ?? [:])
            }
            return
        }
        onNotification?(object)
    }
}
